### PR TITLE
Infrastructure for removing `MATLAB`

### PR DESCRIPTION
**Pull request by samcunliffe**
This PR is a single `git cherry-pick` of the merge commit of #281 (eventually) targeting `main`.

Part of 

* #181 

Continues from
1. #304 
2. #310 
3. #311 
4. #312 

Supercedes
* #193
* #250 

## Actual code notes and change description

... are in 
* #281

### CODE REVIEW

**Review by willGraham01 — APPROVED** [review]

File: tdms/src/input_matrices.cpp [hunk 1]
```diff
@@ -51,6 +52,7 @@ void InputMatrices::set_from_input_file(const char *mat_filename) {
 }
 void InputMatrices::set_from_input_file(const char *mat_filename,
                                         const char *gridfile) {
+  input_filename = std::string(mat_filename);
   MatrixCollection infile_expected(matrixnames_infile);
   MatFileMatrixCollection infile_contains(mat_filename);
   spdlog::info("Input file: " + string(mat_filename) +
```
[the remark this answers is not in the document]
Comment (samcunliffe) [reply tdms/src/input_matrices.cpp:58]: Also, are we `using` the `std` namespace here? I'd vote for not, but if we are, we should probably be consistent.

File: tdms/tests/unit/hdf5_io_tests/test_hdf5_io.cpp [hunk 2]
```diff
@@ -32,7 +27,7 @@ TEST_CASE("Test file I/O construction/destruction.") {
   SECTION("Check file creation.") {
     HDF5Writer f(tmp.string() + "/test_file_constructor.h5");
     CHECK(f.is_ok());
-  }// Destructor called as we leave scope
+  }
```
Comment (samcunliffe) [inline tdms/tests/unit/hdf5_io_tests/test_hdf5_io.cpp:30]:
```suggestion
  }// Destructor called as we leave scope
```

File: tdms/tests/unit/hdf5_io_tests/test_hdf5_io.cpp [hunk 3]
```diff
@@ -61,13 +55,12 @@ TEST_CASE("Test file I/O construction/destruction.") {
...
     HDF5Reader fr(tmp.string() + "/this_file_does_exist_but_is_empty.h5");
-    CHECK_THROWS_AS(fr.read("nonexistantdata", data), H5::FileIException);
+    CHECK_THROWS(fr.read("nonexistantdata", data));
   }
 
   // Normal operation: we should be able to create a file and write to it, then
   // read from it.
   SECTION("Check write then read.") {
-    // Create a file and write some data.
     {
```
Comment (samcunliffe) [inline tdms/tests/unit/hdf5_io_tests/test_hdf5_io.cpp:64]:
```suggestion
    // Create a file and write some data
    {
```

File: tdms/include/hdf5_io/hdf5_dimension.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+#include <H5Cpp.h>
+
+class H5Dimension : public std::vector<hsize_t> {
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply tdms/include/hdf5_io/hdf5_dimension.h:8]: I will leave this like it is for now, however will keep it in mind for when removing our custom `Vector` class (as H5Dimensions can just inherit from that)

File: tdms/src/hdf5_io/hdf5_reader.cpp [hunk 5]
```diff
@@ -1,14 +1,54 @@
...
+  // elements in the buffer.
+  f_vec->x.resize(x_dims.max_dim());
+  f_vec->y.resize(y_dims.max_dim());
+  // Now read the data into the vectors
+  read_dataset_in_group("f_vec", "fx_vec", f_vec->x.data());
+  read_dataset_in_group("f_vec", "fy_vec", f_vec->y.data());
+}
+
+void HDF5Reader::read(Cuboid *cube) const {
+  string cuboid_dataset = "phasorsurface";
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply tdms/src/hdf5_io/hdf5_reader.cpp:39]: Yes. Originally, Peter had planned to allow the user to define a custom surface over which to extract phasors (I believe) but only got as far as implementing this for a cuboid surface.

**Review by samcunliffe — COMMENTED** [review]
🍉

**Pull request merged**